Can't load charts from Chart Library

Some analysts have folios (folders) full of charts (sea maps).

They click on "Insert Chart Library", and choose the folder.  Then a series of red rectangles are plotted, to show where there are maps available.

Analysts then right-click on the corner of a rectangle and select "Load this chart".  The relevant GeoTiff is then loaded.

This is the error:
```
loading:/Volumes/BulkStore/BackedUp/Bulk Data/GeoData - others in git-GeoData/Raster Charts/Gulf_Area_2/38_P1_KARACHI_TO_RA_S_AL_HAD.tif into:The Layers
Jul 03, 2020 4:23:11 PM org.geotools.coverage.grid.io.imageio.MaskOverviewProvider$SpiHelper <init>
WARNING: Unable to create a Reader for File: file:/Volumes/BulkStore/chart_name.tif
java.lang.IllegalArgumentException: XXX
	at com.sun.media.imageioimpl.stream.ChannelImageInputStreamSpi.createInputStreamInstance(ChannelImageInputStreamSpi.java:75)
	at org.geotools.coverage.grid.io.imageio.MaskOverviewProvider$SpiHelper.<init>(MaskOverviewProvider.java:767)
	at org.geotools.coverage.grid.io.imageio.MaskOverviewProvider$SpiHelper.<init>(MaskOverviewProvider.java:748)
	at org.geotools.coverage.grid.io.imageio.MaskOverviewProvider.<init>(MaskOverviewProvider.java:104)
	at org.geotools.coverage.grid.io.imageio.MaskOverviewProvider.<init>(MaskOverviewProvider.java:99)
	at org.geotools.gce.geotiff.GeoTiffReader.getHRInfo(GeoTiffReader.java:362)
	at org.geotools.gce.geotiff.GeoTiffReader.<init>(GeoTiffReader.java:253)
	at org.geotools.gce.geotiff.GeoTiffFormat.getReader(GeoTiffFormat.java:306)
	at org.geotools.gce.geotiff.GeoTiffFormat.getReader(GeoTiffFormat.java:75)
	at org.mwc.cmap.geotools.gt2plot.WorldImageLayer.loadLayer(WorldImageLayer.java:201)
	at org.mwc.cmap.geotools.gt2plot.GeoToolsLayer.setMap(GeoToolsLayer.java:102)
	at org.mwc.cmap.geotools.gt2plot.GtProjection.addGeoToolsLayer(GtProjection.java:215)
	at org.mwc.cmap.plotViewer.editors.CorePlotEditor$3.addThisLayer(CorePlotEditor.java:365)
	at MWC.GUI.Shapes.ChartBoundsWrapper.loadThisChart(ChartBoundsWrapper.java:476)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.mwc.cmap.core.property_support.RightClickSupport$SubjectMethod.run(RightClickSupport.java:256)
	at org.eclipse.jface.action.Action.runWithEvent(Action.java:474)
	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:580)
	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:412)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4364)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1512)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1535)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1520)
	at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1324)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4151)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3768)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1173)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1062)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:635)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:559)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:150)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:155)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:137)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:107)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:400)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:660)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:597)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1468)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1441)

Jul 03, 2020 4:23:11 PM org.geotools.gce.geotiff.GeoTiffFormat getReader
WARNING: java.lang.IllegalArgumentException: XXX
org.geotools.data.DataSourceException: java.lang.IllegalArgumentException: XXX
	at org.geotools.gce.geotiff.GeoTiffReader.<init>(GeoTiffReader.java:266)
	at org.geotools.gce.geotiff.GeoTiffFormat.getReader(GeoTiffFormat.java:306)
	at org.geotools.gce.geotiff.GeoTiffFormat.getReader(GeoTiffFormat.java:75)
	at org.mwc.cmap.geotools.gt2plot.WorldImageLayer.loadLayer(WorldImageLayer.java:201)
	at org.mwc.cmap.geotools.gt2plot.GeoToolsLayer.setMap(GeoToolsLayer.java:102)
	at org.mwc.cmap.geotools.gt2plot.GtProjection.addGeoToolsLayer(GtProjection.java:215)
	at org.mwc.cmap.plotViewer.editors.CorePlotEditor$3.addThisLayer(CorePlotEditor.java:365)
	at MWC.GUI.Shapes.ChartBoundsWrapper.loadThisChart(ChartBoundsWrapper.java:476)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.mwc.cmap.core.property_support.RightClickSupport$SubjectMethod.run(RightClickSupport.java:256)
	at org.eclipse.jface.action.Action.runWithEvent(Action.java:474)
	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:580)
	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:412)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4364)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1512)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1535)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1520)
	at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1324)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4151)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3768)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1173)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1062)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:635)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:559)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:150)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:155)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:137)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:107)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:400)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:660)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:597)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1468)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1441)
Caused by: org.geotools.data.DataSourceException: java.lang.IllegalArgumentException: XXX
	at org.geotools.gce.geotiff.GeoTiffReader.getHRInfo(GeoTiffReader.java:502)
	at org.geotools.gce.geotiff.GeoTiffReader.<init>(GeoTiffReader.java:253)
	... 45 more
Caused by: java.lang.IllegalArgumentException: java.lang.IllegalArgumentException: XXX
	at org.geotools.coverage.grid.io.imageio.MaskOverviewProvider$SpiHelper.<init>(MaskOverviewProvider.java:779)
	at org.geotools.coverage.grid.io.imageio.MaskOverviewProvider$SpiHelper.<init>(MaskOverviewProvider.java:748)
	at org.geotools.coverage.grid.io.imageio.MaskOverviewProvider.<init>(MaskOverviewProvider.java:104)
	at org.geotools.coverage.grid.io.imageio.MaskOverviewProvider.<init>(MaskOverviewProvider.java:99)
	at org.geotools.gce.geotiff.GeoTiffReader.getHRInfo(GeoTiffReader.java:362)
	... 46 more
Caused by: java.lang.IllegalArgumentException: XXX
	at com.sun.media.imageioimpl.stream.ChannelImageInputStreamSpi.createInputStreamInstance(ChannelImageInputStreamSpi.java:75)
	at org.geotools.coverage.grid.io.imageio.MaskOverviewProvider$SpiHelper.<init>(MaskOverviewProvider.java:767)
	... 50 more


```

--- a/org.mwc.cmap.geotools/src/org/mwc/cmap/geotools/gt2plot/GeoToolsLayer.java
+++ b/org.mwc.cmap.geotools/src/org/mwc/cmap/geotools/gt2plot/GeoToolsLayer.java
@@ -16,12 +16,18 @@
 package org.mwc.cmap.geotools.gt2plot;
 
 import java.io.File;
+import java.util.Iterator;
+
+import javax.imageio.spi.IIORegistry;
+import javax.imageio.spi.ImageInputStreamSpi;
 
 import org.geotools.map.Layer;
 import org.geotools.map.MapContent;
 
 import Debrief.GUI.Frames.Application;
 import MWC.GUI.ExternallyManagedDataLayer;
+import MWC.GUI.ToolParent;
+import it.geosolutions.imageio.stream.input.spi.URLImageInputStreamSpi;
 
 public abstract class GeoToolsLayer extends ExternallyManagedDataLayer {
 	/**
@@ -125,4 +131,27 @@ public abstract class GeoToolsLayer extends ExternallyManagedDataLayer {
 			_myLayer.setVisible(visible);
 	}
 
+
+	public static void registerTifUrlServiceProvider() {
+		boolean isRegistered = false;
+		// Ensure that the provider is present
+		try {
+			final Iterator<ImageInputStreamSpi> iter = IIORegistry.getDefaultInstance()
+					.getServiceProviders(ImageInputStreamSpi.class, true);
+
+			while (iter.hasNext() && !isRegistered) {
+				final ImageInputStreamSpi stream = iter.next();
+				if (URLImageInputStreamSpi.class.equals(stream.getClass())) {
+					isRegistered = true;
+				}
+			}
+
+			if (!isRegistered) {
+				IIORegistry.getDefaultInstance().registerServiceProvider(new URLImageInputStreamSpi(),
+						ImageInputStreamSpi.class);
+			}
+		} catch (final IllegalArgumentException e) {
+			Application.logError2(ToolParent.WARNING, "Failure in service registration", e);
+		}
+	}
 }

--- a/org.mwc.cmap.geotools/src/org/mwc/cmap/geotools/gt2plot/GeoToolsLayer.java
+++ b/org.mwc.cmap.geotools/src/org/mwc/cmap/geotools/gt2plot/GeoToolsLayer.java
@@ -37,6 +37,29 @@ public abstract class GeoToolsLayer extends ExternallyManagedDataLayer {
 
 	public static final String NATURAL_EARTH = "Natural Earth";
 
+	public static void registerTifUrlServiceProvider() {
+		boolean isRegistered = false;
+		// Ensure that the provider is present
+		try {
+			final Iterator<ImageInputStreamSpi> iter = IIORegistry.getDefaultInstance()
+					.getServiceProviders(ImageInputStreamSpi.class, true);
+
+			while (iter.hasNext() && !isRegistered) {
+				final ImageInputStreamSpi stream = iter.next();
+				if (URLImageInputStreamSpi.class.equals(stream.getClass())) {
+					isRegistered = true;
+				}
+			}
+
+			if (!isRegistered) {
+				IIORegistry.getDefaultInstance().registerServiceProvider(new URLImageInputStreamSpi(),
+						ImageInputStreamSpi.class);
+			}
+		} catch (final IllegalArgumentException e) {
+			Application.logError2(ToolParent.WARNING, "Failure in service registration", e);
+		}
+	}
+
 	/**
 	 * the map where we display ourselves
 	 *
@@ -129,29 +152,5 @@ public abstract class GeoToolsLayer extends ExternallyManagedDataLayer {
 
 		if (_myLayer != null)
 			_myLayer.setVisible(visible);
-	}
-
-
-	public static void registerTifUrlServiceProvider() {
-		boolean isRegistered = false;
-		// Ensure that the provider is present
-		try {
-			final Iterator<ImageInputStreamSpi> iter = IIORegistry.getDefaultInstance()
-					.getServiceProviders(ImageInputStreamSpi.class, true);
-
-			while (iter.hasNext() && !isRegistered) {
-				final ImageInputStreamSpi stream = iter.next();
-				if (URLImageInputStreamSpi.class.equals(stream.getClass())) {
-					isRegistered = true;
-				}
-			}
-
-			if (!isRegistered) {
-				IIORegistry.getDefaultInstance().registerServiceProvider(new URLImageInputStreamSpi(),
-						ImageInputStreamSpi.class);
-			}
-		} catch (final IllegalArgumentException e) {
-			Application.logError2(ToolParent.WARNING, "Failure in service registration", e);
-		}
 	}
 }

--- a/org.mwc.cmap.geotools/src/org/mwc/cmap/geotools/gt2plot/WorldImageLayer.java
+++ b/org.mwc.cmap.geotools/src/org/mwc/cmap/geotools/gt2plot/WorldImageLayer.java
@@ -198,6 +198,7 @@ public class WorldImageLayer extends GeoToolsLayer {
 		final AbstractGridFormat format = GridFormatFinder.findFormat(openFile);
 		final Hints hints = new Hints(Hints.FORCE_LONGITUDE_FIRST_AXIS_ORDER, Boolean.TRUE);
 
+		GeoToolsLayer.registerTifUrlServiceProvider();
 		tiffReader = format.getReader(openFile, hints);
 
 		/*

--- a/org.mwc.debrief.core/src/org/mwc/debrief/core/loaders/TifLoader.java
+++ b/org.mwc.debrief.core/src/org/mwc/debrief/core/loaders/TifLoader.java
@@ -25,6 +25,7 @@ import javax.imageio.spi.ImageInputStreamSpi;
 import org.eclipse.core.runtime.IAdaptable;
 import org.eclipse.core.runtime.IProgressMonitor;
 import org.eclipse.jface.operation.IRunnableWithProgress;
+import org.mwc.cmap.geotools.gt2plot.GeoToolsLayer;
 
 import Debrief.GUI.Frames.Application;
 import MWC.GUI.ExternallyManagedDataLayer;
@@ -41,7 +42,7 @@ public class TifLoader extends CoreLoader {
 	public TifLoader() {
 		super(".tif", ".tif");
 
-		registerUrlServiceProvider();
+		GeoToolsLayer.registerTifUrlServiceProvider();
 	}
 
 	@Override
@@ -60,28 +61,5 @@ public class TifLoader extends CoreLoader {
 				theLayers.addThisLayer(dl);
 			}
 		};
-	}
-
-	private void registerUrlServiceProvider() {
-		boolean isRegistered = false;
-		// Ensure that the provider is present
-		try {
-			final Iterator<ImageInputStreamSpi> iter = IIORegistry.getDefaultInstance()
-					.getServiceProviders(ImageInputStreamSpi.class, true);
-
-			while (iter.hasNext() && !isRegistered) {
-				final ImageInputStreamSpi stream = iter.next();
-				if (URLImageInputStreamSpi.class.equals(stream.getClass())) {
-					isRegistered = true;
-				}
-			}
-
-			if (!isRegistered) {
-				IIORegistry.getDefaultInstance().registerServiceProvider(new URLImageInputStreamSpi(),
-						ImageInputStreamSpi.class);
-			}
-		} catch (final IllegalArgumentException e) {
-			Application.logError2(ToolParent.WARNING, "Failure in service registration", e);
-		}
 	}
 }

--- a/org.mwc.debrief.core/src/org/mwc/debrief/core/loaders/TifLoader.java
+++ b/org.mwc.debrief.core/src/org/mwc/debrief/core/loaders/TifLoader.java
@@ -17,22 +17,15 @@ package org.mwc.debrief.core.loaders;
 
 import java.io.File;
 import java.io.InputStream;
-import java.util.Iterator;
-
-import javax.imageio.spi.IIORegistry;
-import javax.imageio.spi.ImageInputStreamSpi;
 
 import org.eclipse.core.runtime.IAdaptable;
 import org.eclipse.core.runtime.IProgressMonitor;
 import org.eclipse.jface.operation.IRunnableWithProgress;
 import org.mwc.cmap.geotools.gt2plot.GeoToolsLayer;
 
-import Debrief.GUI.Frames.Application;
 import MWC.GUI.ExternallyManagedDataLayer;
 import MWC.GUI.Layers;
-import MWC.GUI.ToolParent;
 import MWC.GUI.Shapes.ChartBoundsWrapper;
-import it.geosolutions.imageio.stream.input.spi.URLImageInputStreamSpi;
 
 /**
  * @author ian.mayo

